TypeCobol/Compiler/Symbols/Symbol.cs

```diff
@@ -0,0 +1,501 @@
+        /// </summary>
+        public enum Kinds
+        {
+            Global,//The Global Symbol table
+            Namespace,
+            Program,
+            Function,
+            Typedef,
+            Variable,
+            Index,
+            Section,
+            Paragraph
+        }
+
+        [Flags]
+        public enum Flags : ulong
+        {
+            Public = 0x01 << 0,
+            Private = 0x01 << 1,
+            External = 0x01 << 2,
+            Global = 0x01 << 3, //Symbol explicitly marked as global
+            FILE_SECTION = 0x01 << 4,
+            GLOBAL_STORAGE = 0x01 << 5,
+            WORKING_STORAGE = 0x01 << 6,
+            LOCAL_STORAGE = 0x01 << 7,
+            LINKAGE = 0x01 << 8,
+            Input = 0x01 << 9,
+            Output = 0x01 << 10,
+            Inout = 0x01 << 11,
+            ByReference = 0x01 << 12,
+            ByContent = 0x01 << 13,
+            ByValue = 0x01 << 14,
+            Strong = 0x01 << 15,
+            Strict = 0x01 << 16,
+            Weak = 0x01 << 17,
+            Conditions = 0x01 << 18,
+            Renames = 0x01 << 19,
+            Redefines = 0x01 << 20,
+            HasATypedefType = 0x01 << 21,//The symbol has a type that comes from a TypeDef.
+            Parameter = 0x01 << 22,//This a parameter variable.
+            Return = 0x01 << 23,//A Return variable.
+            BuiltinType = 0x01 << 24,//This is a Builtin Type.
+            InsideTypdef = 0x01 << 25,//Flag of any symbol inside a Typedef definition
+
+            //Etc...
+        }
+
+        /// <summary>
+        /// The Visibility mask that a symbol can take.
+        /// </summary>
+        public const Flags SymbolVisibilityMask = Flags.Public | Flags.Private | Flags.External | Flags.Global;
+
+        /// <summary>
+        /// Empty constructor
+        /// </summary>
+        protected Symbol()
+        {
+        }
+
+        /// <summary>
+        /// Named constructor
+        /// </summary>
+        protected Symbol(string name, Kinds kind)
+        {
+            this.Name = name??"";//It happends that a symbol can have no name
+            Kind = kind;
+        }
+
+        /// <summary>
+        /// Symbol Kind
+        /// </summary>
+        public Kinds Kind
+        {
+            get;
+            protected set;
+        }
+
+        /// <summary>
+        /// Symbol Flags.
+        /// </summary>
+        public Flags Flag
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// Symbol's name
+        /// </summary>
+        public string Name
+        {
+            get;
+            internal set;
+        }
+
+        private System.WeakReference _myTargetNode = null;
+        /// <summary>
+        /// The target AST node if any
+        /// </summary>
+        public Node TargetNode
+        {
+            get
+            {
+                lock (this)
+                {
+                    return (Node)_myTargetNode?.Target;
```

Comment:
No check on whether the reference is alive or not ?

Reply:
What reference do you mean ?

Reply:
_myTargetNode is a WeakReference and Target may have been collected when you access it.

Reply:
Yes in this case it will be set to null, this why WeakReference is used, if the Node has been collected, never mind. Symbols are not owner of the node instances. But sure the caller must check the return value against null.